rust/src/util/types.rs

```diff
@@ -1,55 +1,279 @@
+        std::str::from_utf8(self)
+    }
+
+    pub fn as_path(&self) -> Result<PathBuf, Utf8Error> {
+        self.as_str().map(PathBuf::from)
+    }
 }
 
 #[repr(C)]
-#[derive(DropStructMacro)]
-pub struct fil_InitLogFdResponse {
+pub struct fil_Result<T: Sized> {
     pub status_code: FCPResponseStatus,
-    pub error_msg: *const libc::c_char,
+    pub error_msg: fil_Bytes,
+    pub value: T,
 }
 
-impl Default for fil_InitLogFdResponse {
+impl<T: Sized + Default> Default for fil_Result<T> {
     fn default() -> Self {
+        fil_Result {
+            status_code: FCPResponseStatus::FCPNoError,
+            error_msg: Default::default(),
+            value: Default::default(),
+        }
+    }
+}
+
+impl<T: Sized> CodeAndMessage for fil_Result<T> {
+    fn set_error(&mut self, code_and_message: (FCPResponseStatus, *mut libc::c_char)) {
+        let s = unsafe { CStr::from_ptr(code_and_message.1.cast()) }
+            .to_string_lossy()
+            .to_string();
+        self.status_code = code_and_message.0;
+        self.error_msg = s.into();
+    }
+}
+
+impl<T, E> From<Result<T, E>> for fil_Result<T>
+where
+    T: Sized + Default,
+    E: Display,
+{
+    fn from(r: Result<T, E>) -> Self {
+        match r {
+            Ok(value) => Self::ok(value),
+            Err(e) => Self::err(e.to_string()),
+        }
+    }
+}
+
+impl<T> From<T> for fil_Result<T>
+where
+    T: Sized,
+{
+    fn from(value: T) -> Self {
         Self {
-            error_msg: ptr::null(),
             status_code: FCPResponseStatus::FCPNoError,
+            error_msg: Default::default(),
+            value,
+        }
+    }
+}
+
+impl<T: Sized> fil_Result<T> {
+    pub fn ok(value: T) -> Self {
+        fil_Result {
+            status_code: FCPResponseStatus::FCPNoError,
+            error_msg: Default::default(),
+            value,
+        }
+    }
+
+    pub unsafe fn into_boxed_raw(self) -> *mut fil_Result<T> {
+        Box::into_raw(Box::new(self))
+    }
+}
+
+impl<T: Sized + Default> fil_Result<T> {
+    pub fn err(err: impl Into<fil_Bytes>) -> Self {
+        fil_Result {
+            status_code: FCPResponseStatus::FCPUnclassifiedError,
+            error_msg: err.into(),
+            value: Default::default(),
         }
     }
 }
 
-code_and_message_impl!(fil_InitLogFdResponse);
+#[allow(non_camel_case_types)]
+pub type fil_GpuDeviceResponse = fil_Result<fil_Array<fil_Bytes>>;
+
+#[no_mangle]
+pub unsafe extern "C" fn fil_destroy_gpu_device_response(ptr: *mut fil_GpuDeviceResponse) {
+    let _ = Box::from_raw(ptr);
+}
+
+#[allow(non_camel_case_types)]
+pub type fil_InitLogFdResponse = fil_Result<()>;
 
 #[no_mangle]
 pub unsafe extern "C" fn fil_destroy_init_log_fd_response(ptr: *mut fil_InitLogFdResponse) {
     let _ = Box::from_raw(ptr);
 }
+
+pub fn vec_into_raw<T>(v: Vec<T>) -> (*mut T, usize) {
+    let bytes = v.into_boxed_slice();
+    let len = bytes.len();
+    (Box::into_raw(bytes).cast(), len)
```

Comment:
So... if this vec is empty, the pointer will be a pointer to `1`, not null. If we pass that to _go_, and try to construct a slice from that, go will segfault (speaking from experience).

Reply:
removed